src/cpp/server/ServerPAMAuth.cpp

```diff
@@ -147,7 +149,39 @@ std::string getUserIdentifier(const core::http::Request& request)
 
 std::string userIdentifierToLocalUsername(const std::string& userIdentifier)
 {
-   return userIdentifier;
+   static core::thread::ThreadsafeMap<std::string, std::string> cache;
+   std::string username = userIdentifier;
+
+   if (cache.contains(userIdentifier)) 
+   {
+      username = cache.get(userIdentifier);
+   }
+   else
+   {
+      // The username returned from this function is eventually used to create
+      // a local stream path, so it's important that it agree with the system
+      // view of the username (as that's what the session uses to form the
+      // stream path), which is why we do a username => username transform
+      // here. See case 5413 for details.
+      core::system::user::User user;
+      Error error = core::system::user::userFromUsername(userIdentifier, &user);
+      if (error) 
+      {
+         // log the error and return the PAM user identifier as a fallback
+         LOG_ERROR(error);
```

Comment:
Yes, that's the intent! After this error is logged we go ahead and cache the username on L181, so it shouldn't be logged more than once per user (per session) in the worst case.